sound/soc/sof/ipc4-pcm.c

```diff
@@ -495,17 +519,51 @@ static int sof_ipc4_pcm_dai_link_fixup(struct snd_soc_pcm_runtime *rtd,
 				use_chain_dma = true;
 		}
 	}
+
+	fe_rate = params_rate(params);
+
 	/*
-	 * Always set BE format to 32-bits for both playback and capture, except if chained
-	 * DMA is in use. In chained DMA case the back-end and front-end format must match.
+	 * Set audio format and rate mask based on available formats
+	 * defined for the DAI/BE copier. Copiers are not used with
+	 * chain_dma, so must be skipped for this case. Copiers can
+	 * handle arbitrary number of channels, so channel count is
+	 * not limited here.
 	 */
 	if (!use_chain_dma) {
-		snd_mask_none(fmt);
-		snd_mask_set_format(fmt, SNDRV_PCM_FORMAT_S32_LE);
+		struct sof_ipc4_available_audio_format *available_fmt = &ipc4_copier->available_fmt;
+		int i;
+
+		snd_mask_none(fmt_mask);
+
+		for (i = 0; i < available_fmt->audio_fmt_num; i++) {
+			struct sof_ipc4_audio_format *fmt =
+				&available_fmt->base_config[i].audio_fmt;
+			unsigned int bit_depth;
+
+			bit_depth = SOF_IPC4_AUDIO_FORMAT_CFG_V_BIT_DEPTH(fmt->fmt_cfg);
+			if (ipc4_set_fmt_mask(fmt_mask, bit_depth) < 0)
+				return -EINVAL;
+
+			fmt = &available_fmt->out_audio_fmt[i];
+			bit_depth = SOF_IPC4_AUDIO_FORMAT_CFG_V_BIT_DEPTH(fmt->fmt_cfg);
+			if (ipc4_set_fmt_mask(fmt_mask, fmt->bit_depth) < 0)
+				return -EINVAL;
```

Comment:
@kv2019i I am having a hard time understanding the logic with the loop here. For every available format, you first set the format mask based on the input format and then override it with the output format. Can you please add some explanation as to what we're trying to do here?

Reply:
@ranj063 The "set" verb is misleading but I'm sticking to ALSA semantics where "snd_mask_set()" actually does a logical OR to the mask, i.e. formats are incrementally added to the mask.

I do agree this is still confusing. This code is in dai_link_fixup, and the dailink is really bidirectional. What we want here is to limit the DAI link to the formats declared in the topology. Unfortunately it is possible to have different formats as copier inputs and outputs. Here I'm declaring all formats (both at input and outputs). This defines a superset (e.g. both S16 and S32), even if for particular direction (like capture) only a a subset is actually supported (S32).

This still seems improvement over existing code where we just hardcode to S32 with no input from topology.

But, but, one option I'm considering whether it would be better not to limit the formats in dai_link_fixup at all. Having the format negotiation done in the path enabling code (e.g. in copier_prepare), might be sufficient and having this in one place is probably easier to understand.

In IPC3 code we do look up more constraints from tplg (e.g. format from private->comp_dai->config.frame_fmt) and set them in dai_link_fixup. And thus I tried to do the same for IPC4.